netperf_tool/netperf_tool/__init__.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from inspect import isclass
+
+import rclpy.qos
+from rpyutils import add_dll_directories_from_env
+
+# Since Python 3.8, on Windows we should ensure DLL directories are explicitly added
+# to the search path.
+# See https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
+with add_dll_directories_from_env('PATH'):
+    from netperf_tool.netperf_tool_impl import (
+        ClientRunner as _ClientRunner,
+        ServerRunner as _ServerRunner,
+    )
+
+
+class RunnerImpl:
+    _impl_cls = None
+
+    def __init__(
+        self,
+        experiment_duration: float,
+        qos: rclpy.qos.QoSProfile,
+        *args
+    ):
+        if self._impl_cls is None or not isclass(self._impl_cls):
+            raise RuntimeError('internal error')
+        self._impl = self._impl_cls(qos.get_c_qos_profile(), *args)
+        self._experiment_duration = experiment_duration
+
+    def __enter__(self):
+        self._impl.start(self._experiment_duration)
+        return self._impl.get_node()
+
+    def stop(self):
+        self._impl.stop()
+
+    def wait_for_experiment_to_complete(self):
+        self._impl.join()
+
+    def __exit__(self, _, _1, _2):
+        self._impl.stop()
+        self._impl.join()
+
+
+class ClientRunner(RunnerImpl):
+    _impl_cls = _ClientRunner
+
+    def __init__(
+        self,
+        *,
+        experiment_duration_s: float,
+        qos: rclpy.qos.QoSProfile,
+        message_size_bytes: int,
+        target_pub_period_s: float
+    ):
+        super().__init__(
+            experiment_duration_s, qos, message_size_bytes, target_pub_period_s)
+
+
+class ServerRunner(RunnerImpl):
+    _impl_cls = _ServerRunner
+
+    def __init__(
+        self,
+        *,
+        qos: rclpy.qos.QoSProfile
+    ):
+        super().__init__(-1e-9, qos)
```

Comment:
This is an odd magic number, maybe it needs a comment or should be a constant. I can't really tell what this is supposed to mean (just looking here).

Reply:
Sorry about that, I only need a negative number really.
The python value is a floating point number in seconds, which is converted to `std::chrono::nanoseconds`.
In this case, `-1e9` is `-1ns` in python.
I could add a comment, or instead use `-1.`, which is less confusing.

See https://pybind11.readthedocs.io/en/stable/advanced/cast/chrono.html#provided-conversions.

Reply:
I could maybe use `datetime.timedelta` instead of floating point numbers to avoid confusion as well.

Reply:
Whatever you do, it's not clear from this file why it's a negative number.

What not just `-1.0` or even better a constant like `BLOCK_FOREVER`?

Reply:
Done, see https://github.com/ros2/ros2cli/pull/720/commits/c14657702684f996e694ee718896272250e409f8